packages/nimble-components/src/wafer-map/specs/README.md

```diff
@@ -0,0 +1,199 @@
+
+Please see below a few screenshots regarding the current `wafer-map` component:
+
+![WaferMap Example](file:///Resources/wafer_map_example.png)
+
+## Design
+
+From a user’s perspective one of the most important actions which this component should be able to achieve is to zoom in and out on the datapoints. This functionality should occur when the mouse pointer hovers the `nimble-wafer-map` component and a mouse wheel action is performed.
+While zooming in on the component, the user should also be able to pan (drag) the wafer map by holding the left mouse button and moving the cursor on the XY axis.
+
+From a developer perspective, it's very important that the `nimble-wafer-map` component can be integrated with the MicroStrategy Custom plugin framework as this framework provides functionalities as follows:
+* Tooltip when hovering any if the datapoints withing the wafer map
+* Wafer map legend
+* MicroStrategy specific context menus on right-click action
+
+Information regarding the basic MicrosStrategy **Custom Plugin Workflow** can be found [here](https://www2.microstrategy.com/producthelp/Current/VisSDK/Content/topics/HTML5/Understanding_the_workflow.htm)
+Also, the visualisation widget has no active controls, so there is no way to alter the visual aspect or data post rendering. Every input parameter must be provided through the constructor of the wafer map component.
+
+### API
+
+*The key elements of the component's public API surface:*
+* Component Name: `nimble-wafer-map`
+* Props/Attrs: 
+  * `die[]` - this represents the input data, which fills the wafer map with content.\
+  The **die**  object contains the following attributes:
+	* XCoorrdianate: integer
+	* YCoordinate: integer
+	* Value: float
+  * `quadrant` - represents the orientation fo the dies on the wafer map (the layout of the values on the dies). It can be represented by an Enum with the following values: 
+	* TopLeft - ![Top Left Quadrant](file:///Resources/top_left.png)
+	* BottomLeft - ![Bottom Left Quadrant](file:///Resources/bottom_left.png)
+	* TopRight - ![Top Right Quadrant](file:///Resources/top_right.png)
+	* BottomRight - ![Bottom Right Quadrant](file:///Resources/bottom_right.png)
+  * `orientation` - represent the orientation of the notch on the wafer map outline (only visual). As only four static orientations are possible, it can be represented by an Enum with the following values: top, bottom, left, right.
+  * `color_scale` - represents the color spectrum which shows the status of the dies on the wafer.
+  * `number_of_characters` - represents the number of characters allowed to be displayed within a single die. As the die values are represented by Floating point numbers, we must have the liberty of limiting how many characters we are willing to display within a single die.
+  
+ Please note that all these attributes are only accessible through the constructor of the component. Once set, the only way to modify them is to re-render the component.
```

Comment:
Just to think through some of the consequences of this:
1. there is no attribute API for this component (e.g. I can't set `<nimble-wafer-map orientation='right'>`
2. the property API is read-only (e.g. in JS code I can read `waferMap.orientation` but not write it)
3. there is no way to update the data in the wafer map other than constructing a new wafer map (I can't set `waferMap.dies = [...];` or call a method like `refresh()`/`render()`

If my understanding is correct then I'd propose rewriting like this:

```suggestion
 Please note that all these properties are only settable through the constructor of the component, not via properties or attributes. Once set, the only way to modify them is to initialize a new component.
```

Reply:
I'd like to dig into this a bit more from a reasoning perspective. Why shouldn't the API be settable? What makes a wafer-map distinct enough in its usage, from other data visualization components like graphs, that suggest that a user should not be able to modify its state after construction? At the moment, being somewhat naive, I don't see the benefit.

Reply:
> Please note that all these attributes are only accessible through the constructor of the component. Once set, the only way to modify them is to re-render the component.

Custom element APIs are a bit different in that you cannot pass properties to a constructor for an element. The browser owns the constructor and calls it internally, we can't pass options to it.

For example, if I have the HTML in a page of `<input type="button">` that is equivalent to the code:

```ts
const el = document.createElement('input');
input.type = 'button';
```

So behaving like a DOM element is a **requirement** for a custom element based API. Configuration has to be via writeable properties and attribute.

**This should not be a major concern.** The FAST library we build on gives us [internal observability behaviors](https://www.fast.design/docs/fast-element/observables-and-state/#internal-observation) to make updates straightforward.

Roughly the approach would be to create attributes, listen for attribute changes for properties that require recreating re-rendering, and re-rendering the internal state on a `requestAnimationFrame`.


Some pseduocode:
```ts
class WaferMap extends FastElement {
    private renderQueued = false;

    @attr
    orientation;

    @attr
    quadrant;

    orientationChanged() {
       this.queueRender()
    }

   quadrantChanged() {
       this.queueRender()
   }

   private queueRender() {
      if (!this.renderQueued) {
          this.renderQueued = true;
          window.requestAnimationFrame(() => {this.render()})
      }
   }
   
   private render () {
        this.renderQueued = false;
        // Re-render the component
   }
}

```

So some insights are because we would render on requestAnimationFrame this gives us some benefits:

- We only do work for re-rendering the element as the browser draws frames on the screen
- If the user rapidly changes settings we do not waste time doing unnecessary work before the component is rendered again to draw on the screen

So the spec language I would use would be something like:

```suggestion
    As the wafer map will use the d3 library to draw it's contents, attribute value changes will cause the component to re-render on updates. To avoid unnecessary re-rendering from rapid attribute updates, the component will re-render its content during `requestAnimationFrame`.
```

Reply:
Thanks @rajsite. I had been dissuading the team from exposing an attribute API because so many of the properties are complex data types that would have to be encoded/decoded to (presumably JSON) strings. Do you think that's a compelling argument to only expose a property API or is there still value in exposing an attribute API even if it requires some slightly clunky data conversion? I don't feel strongly so could be easily swayed.

@atmgrifter00's question about exposing a settable API is probably moot given the inability to set data in the constructor. But I'm still interested in the original rationale. Is it just that current use cases don't require updating the wafer map after creation or is there a technical challenge to doing so.

Reply:
@jattasNI @zszilagy 

I'd advocate that:
 - the simple properties (numbers, boolean, enums, strings) should be configurable via properties and attributes
   - quadrant, orientation, dieCharacterCount, disabled
 - the complex properties should only be configurable via properties and have no attribute
   - die, and colorScale

And based on the recent push I would advocate we keep the following methods:
- focus()

But remove the following methods:
- setWaferMapOrientation(): orientation should be a writeable property and attribute as mentioned above
-  setWaferMapQuadrant(): quadrant should be a writeable property as mentioned above
- render() should not be a method, the above properties, attributes, and methods should queue rendering as needed

Also @zszilagy you may need to click the `load more` button if you intended to respond to this thread:

![image](https://user-images.githubusercontent.com/1588923/201205640-58642261-7411-46dd-90a7-cf63b035019a.png)

Reply:
@rajsite Thanks for the suggestions. I included them into the spec and will resolve this thread once I fixed everything.